src/app/SimpleSubscriptionResumptionStorage.h

```diff
@@ -0,0 +1,147 @@
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP CASE Session object that provides
+ *      APIs for constructing a secure session using a certificate from the device's
+ *      operational credentials.
+ */
+
+#pragma once
+
+#include <app/SubscriptionResumptionStorage.h>
+#include <lib/core/TLV.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * An example SubscriptionResumptionStorage using PersistentStorageDelegate as it backend.
+ */
+class SimpleSubscriptionResumptionStorage : public SubscriptionResumptionStorage
+{
+public:
+    CHIP_ERROR Init(PersistentStorageDelegate * storage)
+    {
+        VerifyOrReturnError(storage != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+        mStorage = storage;
+        return CHIP_NO_ERROR;
+    }
+
+    static StorageKeyName GetStorageKey(const ScopedNodeId & node);
+
+    CHIP_ERROR LoadIndex(SubscriptionIndex & index) override;
+
+    CHIP_ERROR FindByScopedNodeId(ScopedNodeId node, SubscriptionList & subscriptions) override;
+
+    CHIP_ERROR Save(SubscriptionInfo & subscriptionInfo) override;
+
+    CHIP_ERROR Delete(const SubscriptionInfo & subscriptionInfo) override;
+
+    CHIP_ERROR DeleteAll(FabricIndex fabricIndex) override;
+
+private:
+    CHIP_ERROR SaveIndex(const SubscriptionIndex & index);
+    CHIP_ERROR SaveSubscriptions(const ScopedNodeId & node, const SubscriptionList & subscriptions);
+
+    static constexpr size_t MaxScopedNodeIdSize() { return TLV::EstimateStructOverhead(sizeof(NodeId), sizeof(FabricIndex)); }
+
+    static constexpr size_t MaxIndexSize()
+    {
+        // The max size of the list is (1 byte control + bytes for actual value) times max number of list items
+        return TLV::EstimateStructOverhead((1 + MaxScopedNodeIdSize()) * CHIP_IM_MAX_NUM_SUBSCRIPTIONS);
+    }
+
+    static constexpr size_t MaxSubscriptionPathsSize()
+    {
+        return TLV::EstimateStructOverhead(
+            TLV::EstimateStructOverhead(sizeof(uint8_t), sizeof(EndpointId), sizeof(ClusterId), sizeof(AttributeId)) *
+            CHIP_IM_MAX_NUM_PATH_PER_SUBSCRIPTION);
+    }
+
+    static constexpr size_t MaxSubscriptionSize()
+    {
+        // All the fields added together
+        return TLV::EstimateStructOverhead(MaxScopedNodeIdSize(), sizeof(SubscriptionId), sizeof(uint16_t), sizeof(uint16_t),
```

Comment:
Consider adding slack to account for future fields so that on a downgrade/upgrade, you don't get into a situation of errors that follow-through the top. Another way to deal is to make sure that on deserialization errors, for the subs/module, you *delete* the keys rather than funnel the error all the way up to the top

Reply:
That's a good idea. I'll delete keys on deserialization.

Reply:
If you do so, the sub will not persist since there will be no further save. Actually, what happens today? If you call `ResumeSubscriptions()` at init, do the subs get re-saved?

Reply:
Subs are persisted and retried on reboot. They get deleted on Subscriber responding error or if CASE session fails to establish in the current implementation.

Reply:
And also I've added logic to delete keys on deserialization errors.